### ISSUE TITLE
Pull requests list sometimes doesn't update, even after restarting multiple times and fetching

### ISSUE THREAD

**Issue by mattnewport** [desktop/desktop#5865]
## Description
The "Pull requests" tab on the "Current branch" dropdown frequently does not update with the latest PRs visible on the GitHub website. Sometimes restarting GitHub Desktop resolves this but not always, and it shouldn't be necessary to restart anyway to get an up to date list of PRs.

## Version

* GitHub Desktop: 1.4.2

* Operating system: Windows 10

## Steps to Reproduce

1. Work in a repo with other users, ask them to create a PR.
2. Observe that the PR is visible on the GitHub website.
3. Observe that the PR is sometimes not visible in GitHub Desktop "Current branch" dropdown, even though it is visible on the website.
4. Observe that restarting GitHub Desktop sometimes results in the PR list refreshing but sometimes it doesn't.

### Expected Behavior

After clicking "Fetch origin", any PRs that are visible on the GitHub website should be visible in the "Current branch" dropdown on the "Pull requests" tab.

### Actual Behavior

Sometimes PRs are visible on the website but will not show in GitHub Desktop for an extended period of time despite restarting and fetching multiple times. Sometimes a restart will fix the issue when a fetch does not.

## Additional Information

![image](https://user-images.githubusercontent.com/1325985/46771491-96b7cc80-cca8-11e8-8fd9-6a7f39478eea.png)

![image](https://user-images.githubusercontent.com/1325985/46771509-b4853180-cca8-11e8-8f92-38e0af618c18.png)

### Logs

[2018-10-10.desktop.production.log](https://github.com/desktop/desktop/files/2466788/2018-10-10.desktop.production.log)

**Comment by steveward** [comment]
Thanks for contributing @mattnewport. I've been trying to reproduce this behavior, but have so far been unsuccessful. Is this happening in multiple repositories, or just one specific repository?

I do see some fetch failures in the log file related to one specific repository:

```
2018-10-10T21:12:23.145Z - info: [ui] Executing fetch: git -c credential.helper= fetch --progress --prune origin (took 1.408s)
2018-10-10T21:12:23.520Z - info: [ui] skipping fast-forward for all branches as there are 35 local branches - this will run again when there are less than 20 local branches tracking remotes
2018-10-10T21:20:51.857Z - info: [ui] Background fetch for 9 repositories took 2.553sec
2018-10-10T21:35:51.870Z - info: [ui] Background fetch for 9 repositories took 2.567sec
2018-10-10T21:39:03.177Z - warn: [ui] getFetchPollInterval: failed for ossovr/Osso-VR
TypeError: Failed to fetch
```

Is that possibly the repository where you are experiencing this issue?

**Comment by mattnewport** [comment]
Yes, that looks like the repo that is having the problem. It's the repo for our company's main product so it's heavily used (we have a team of around 10 who at least sometimes commit to it, though a smaller number do so daily).

Sent with [ProtonMail](https://protonmail.com) Secure Email.

‐‐‐‐‐‐‐ Original Message ‐‐‐‐‐‐‐

**Comment by shiftkey** [comment]
I'm going to close this out as we've revamped our Pull Request list for the 2.0 release to cache better and be more responsive to changes.

Please read along with these PRs for more information: #7501 #7163 #7027

**Comment by dpurvis3** [comment]
Same issue in 2.2.4

If you're feeling desperate, you can blow away your `%AppData%\GitHub Desktop` folder.

Solved my problem! 💣😉

**Comment by jzabroski** [comment]
I also ran into this issue.  I am opening an issue because I am very disappointed to see this was force closed with the reasoning "we revamped (so please test for us)".

**Comment by jzabroski** [comment]
Here is a more succinct workaround, based on steps provided by @dproviden:

1. CLOSE GitHub Desktop application
2. Go to `%AppData%/GitHub Desktop`
3. Delete IndexedDB folder
4. Re-start GitHub Desktop application
5. Go to `File -> Add local repository... (Control+O)`
6. Add the repository with the issue back into your local database by selecting it in the file explorer.

This works, but is annoying.

**Comment by billygriffin** [comment]
Hi folks, sorry you're all experiencing this problem. 😞 We've tried reproducing this using several different configurations, and I tried the steps described by @jzabroski in the other issue to no avail.

<img width="650" alt="Screen Shot 2020-01-28 at 1 56 08 PM" src="https://user-images.githubusercontent.com/5091167/73304785-23580800-41d6-11ea-9c77-c97bdf696a1e.png">

Is there anything specific about your setup that might be causing the issue? And similarly if others are running into problems, we'd love to hear what OS you're using and anything else that might explain why it's only happening for a select few folks.

**Comment by jzabroski** [comment]
@billygriffin Are you saying _you cannot reproduce the cache header being sent when I click the Refresh button_?

**Comment by jzabroski** [comment]
@lee-dohm I find it hard to believe this cannot be reproduced given my instructions.  I even screenshotted the root cause: the cache header is being sent when you click Refresh.  I also find it very hard to believe this can't be reproduced after reading the developer explain all the issues (#7501 #7163 #7027) related to even getting this working remotely properly.

Sorry if my comments violated Code of Conduct, but I was frustrated to be told to go read three separate issues for all the work done to make this work.  It seems to me like the PRs API and/or GitHub Desktop downloading of the data is buggy.  I can't really make sense of what's going on but I've decided to stop using GitHub Desktop altogether.

I suggest this ticket be left open, and someone do a deep dive on this, as I think the behavior is likely very wrong in very scary ways. - You may even want to rip this feature out of the product entirely based on other stuff I've observed.  For example, if a PR is updated, it's not clear to me this tab is pulling in those updates!

**Comment by jzabroski** [comment]
@billygriffin I just uninstalled and reinstalled GitHub for Desktop and was able to immediately reproduce the issue on a new repository.  Please elaborate when you say "We've tried reproducing this using several different configurations" what configurations you're referring to.

**Comment by dproviden** [comment]
> Is there anything specific about your setup that might be causing the issue? And similarly if others are running into problems, we'd love to hear what OS you're using and anything else that might explain why it's only happening for a select few folks.

@billygriffin I'm on Windows 10, connecting via HTTPS to a GitHub Enterprise Server. That's all I've got at the moment 🙂

**Comment by niik** [comment]
Hey all, sorry to hear you're having problems with the PR list. I have some follow-up questions that I hope can help us narrow down the potential root cause here. As @billygriffin already stated we haven't been able to reproduce this issue ourselves which could be for any number of reasons but all of us on the Desktop team do use the in-app PR list in our day to day work without experiencing this problem so we need to figure out how our usage differs from yours.

 1. Have you ever had more than one copy of the repository open in Desktop?
  In other words, have you ever cloned the offending repository to more than one location on disk and added both locations to the application?
 2. Do you ever work in a fork of the offending repository?

If you're able to reproduce this problem we'd love it if you could help us troubleshoot this by following the steps below.

1. Open DevTools (View -> Toggle Developer Tools).
2. Click on the Network tab
3. Click the refresh button in the Pull Request list
  ![image](https://user-images.githubusercontent.com/634063/73551688-29a0eb00-4447-11ea-964b-44788f94ace2.png)
4. Find the corresponding request in the network tab and click on it
  ![image](https://user-images.githubusercontent.com/634063/73551735-40dfd880-4447-11ea-942e-55b15372824a.png)
5. Click on the preview tab and expand the first pull request in the list and make a note of the `updated_at` field (or take a screenshot of the whole thing if you're comfortable sharing the PR details)
  ![image](https://user-images.githubusercontent.com/634063/73551830-708ee080-4447-11ea-890e-14592351c96c.png)
6. Go to the Application Tab in DevTools and make your way to `Storage -> IndexedDB -> Database -> repositories -> path` in the sidebar and find the row corresponding to the repository you're having problems with in the right hand pane and make note of its `Primary key`
![image](https://user-images.githubusercontent.com/634063/73552018-bcda2080-4447-11ea-91e5-fe6c6e65459d.png)
7. Now locate the `PullRequestDatabase -> pullRequestsLastUpdated` table in the sidebar and find the row where `Key` corresponds to the `Primary key` you noted in the previous step and make a note of the `lastUpdated` property for that row.
![image](https://user-images.githubusercontent.com/634063/73552170-088cca00-4448-11ea-9758-a0006f3119f0.png)
 
Sharing the results of the above would go a long way towards helping us understand what's going on.

**Comment by mattnewport** [comment]
I do have multiple copies of the repository where I was seeing the problem and they would all have been open in desktop. I'm not using desktop much these days as I've mostly switched to using magit but if I do happen to repro the issue I'll try and send the info.

**Comment by jzabroski** [comment]
> * Have you ever had more than one copy of the repository open in Desktop?
>   In other words, have you ever cloned the offending repository to more than one location on disk and added both locations to the application?
> * Do you ever work in a fork of the offending repository?

Yes to both.

For RazorLight, before I was a co-maintainer, I was a contributor, so I had my own fork and _since the project was unmaintained for a long time, I had forked other people's forks, as well_, in an attempt to herd the cats.  (Note: I don't think GitHub handles forking other people's forks particularly well.  If it does, I am at a loss as to how it does it due to the namespace shadowing issue with GitHub URLs assuming that username/ForkedProject is acceptable.  It would have probably have been better URL structure to have username/ForkedProjectUserName/ForkedProject even if longer to type.

The OTHER weird thing I can share is that, ~6 hours later, without doing anything, I went back to the app and all of a sudden the PRs were there.  That was even weirder.

Again, worth noting the logs say nothing about this, so not sure how else to help reproduce.

**Comment by billygriffin** [comment]
Thanks @mattnewport and @jzabroski!

@jzabroski Did you happen to update to the latest version, 2.3.1? The work we done relating to forks and PRs there *may* have fixed this issue even in the edge cases where it has manifested, so I'm curious if it's still happening when you're on that version.

**Comment by jzabroski** [comment]
I did not upgrade. When I reported my issue, the program told me I was on
the latest version. I will try on Monday

**Comment by dproviden** [comment]
Happened to run into this issue this morning. Here are my screenshots:

#### Network Tab after PR Refresh
![image](https://user-images.githubusercontent.com/19540349/74159209-64deaf00-4be9-11ea-847a-45824083d75f.png)

#### PullRequestDatabase.pullRequestsLastUpdated
![image](https://user-images.githubusercontent.com/19540349/74159343-9e171f00-4be9-11ea-9e9a-3c9cfdd26d99.png)

**Comment by dproviden** [comment]
Also potentially relevant - just switched my authentication from passphrase-protected ssh key to an unprotected ssh key, and now I see my PR.

**Comment by jzabroski** [comment]
> I did not upgrade. When I reported my issue, the program told me I was on the latest version. I will try on Monday

@billygriffin I was indeed on the latest version when I reported the issue, and still am.  Sorry, I was confused by your question - I thought you put out a hotfix after I opened #8995

Note, it's a bit weird #8995 was closed and yet #5865 remains closed, too.  How is this being tracked?  As an independent user, I can't re-open someone elses issue.  It's a bit frustrating when there are new comments in Closed issues, as I wonder if these get surfaced to managers on a project so they are aware a problem is not resolved / customers aren't satisfied.

![image](https://user-images.githubusercontent.com/447485/74193097-c96b2f80-4c24-11ea-86b4-af2e1d3721ad.png)

**Comment by jzabroski** [comment]
This feature is really buggy.  Randomly, the UI adds Pull Requests as Branches in the UI, but it doesn't consistently do this.  

Clicking the PR does not seem to do anything productive:

![image](https://user-images.githubusercontent.com/447485/74257836-62508800-4cc3-11ea-9a02-bdbb452b0dc0.png)

![image](https://user-images.githubusercontent.com/447485/74257949-8a3feb80-4cc3-11ea-8e66-ddb90824d80a.png)
